File: card/easy/others/002.go
package others

// Hamming Distance
// https://leetcode.com/explore/interview/card/top-interview-questions-easy/99/others/762/
// https://leetcode.com/problems/hamming-distance/description/

func hammingDistance(x int, y int) int {
	xor := x ^ y
	distance := 0
	for xor != 0 {
		distance += xor & 0x01
		xor >>= 1
	}
	return distance
}

File: card/easy/sortingandsearching/002/002.cpp
#include <stdio.h>

// First Bad Version
// https://leetcode.com/explore/interview/card/top-interview-questions-easy/96/sorting-and-searching/774/
// https://leetcode.com/problems/first-bad-version/description/

// Forward declaration of isBadVersion API.

bool isBadVersion(int version) {
    // return version >= 4;
    return version >= 1702766719;
}

int firstBadVersion(int n) {
    int left = 1;
    int right = n;
    
    int mid = 0;
    int first = 0;
    while(left <= right){
        mid = ((long)left + right) / 2;
        if (isBadVersion(mid)) {
            first = mid;
            right = mid - 1;
        } else {
            left = mid + 1;
        }   
    }
    return first;
}

int main(int argc, char const *argv[])
{

    printf("%d\n", firstBadVersion(2126753390));
    return 0;
}
